sentry-apollo/src/main/java/io/sentry/apollo/SentryApolloInterceptor.kt

```diff
@@ -0,0 +1,105 @@
+    private val beforeSpan: BeforeSpanCallback? = null
+) : ApolloInterceptor {
+
+    constructor(hub: IHub) : this(hub, null)
+    constructor(beforeSpan: BeforeSpanCallback) : this(HubAdapter.getInstance(), beforeSpan)
+
+    override fun interceptAsync(request: InterceptorRequest, chain: ApolloInterceptorChain, dispatcher: Executor, callBack: CallBack) {
+        val activeSpan = hub.span
+        if (activeSpan == null) {
+            chain.proceedAsync(request, dispatcher, callBack)
+        } else {
+            val span = startChild(request, activeSpan)
+            val sentryTraceHeader = span.toSentryTrace()
+
+            // we have no access to URI, no way to verify tracing origins
+            val headers = request.requestHeaders.toBuilder().addHeader(sentryTraceHeader.name, sentryTraceHeader.value).build()
+            val requestWithHeader = request.toBuilder().requestHeaders(headers).build()
+            span.setTag("operationId", requestWithHeader.operation.operationId())
+            span.setTag("variables", requestWithHeader.operation.variables().valueMap().toString())
+
+            chain.proceedAsync(requestWithHeader, dispatcher, object : CallBack {
+                override fun onResponse(response: InterceptorResponse) {
+                    span.status = response.httpResponse.map { SpanStatus.fromHttpStatusCode(it.code(), SpanStatus.INTERNAL_ERROR) }
+                        .or(SpanStatus.UNKNOWN)
+
+                    finish(span, requestWithHeader, response)
+                    callBack.onResponse(response)
+                }
+
+                override fun onFetch(sourceType: FetchSourceType) {
+                    callBack.onFetch(sourceType)
+                }
+
+                override fun onFailure(e: ApolloException) {
+                    span.apply {
+                        status = SpanStatus.INTERNAL_ERROR
+                        throwable = e
+                    }
+                    finish(span, requestWithHeader)
+                    callBack.onFailure(e)
+                }
+
+                override fun onCompleted() {
+                    callBack.onCompleted()
+                }
+            })
+        }
+    }
+
+    override fun dispose() {}
+
+    private fun startChild(request: InterceptorRequest, activeSpan: ISpan): ISpan {
+        val operation = request.operation.name().name()
+        val operationType = when (request.operation) {
+            is Query -> "query"
+            is Mutation -> "mutation"
+            is Subscription -> "subscription"
+            else -> request.operation.javaClass.simpleName
+        }
+        val description = "$operationType $operation"
+        return activeSpan.startChild(operation, description)
+    }
+
+    private fun finish(span: ISpan, request: InterceptorRequest, response: InterceptorResponse? = null) {
+        var newSpan: ISpan = span
+        if (beforeSpan != null) {
+            newSpan = beforeSpan.execute(span, request, response)
```

Comment:
there's no chance of `execute` throwing then? Since there's `onFailure` ?

Reply:
it can throw if the clients callback code throws, if this is an issue we should handle, we have to fix it on the other integrations too (okhttp, feign and this one), good catch